ryhope/src/tree/sbbst.rs

```diff
@@ -300,11 +344,70 @@ impl State {
             None
         }
     }
+
+    fn inner_idx(&self, outer_idx: OuterIdx) -> InnerIdx {
+        InnerIdx(outer_idx.0 - self.shift)
+    }
+
+    fn outer_idx(&self, inner_idx: InnerIdx) -> OuterIdx {
+        OuterIdx((inner_idx + self.shift).0)
+    }
+}
+
+trait IndexMapper: Sized + Send + Sync + Clone {
+    fn to_inner_idx(&self, outer_idx: OuterIdx) -> impl Future<Output = InnerIdx> + Send;
+
+    fn to_outer_idx(&self, inner_idx: InnerIdx) -> impl Future<Output = OuterIdx> + Send;
+
+    /// Apply `to_outer_idx` to `inner_idx` if it is `Some`, otherwise return `None`.
+    fn to_outer_idx_map(
+        &self,
+        inner_idx: Option<InnerIdx>,
+    ) -> impl Future<Output = Option<OuterIdx>> + Send {
+        async move {
+            match inner_idx {
+                Some(inner_idx) => Some(self.to_outer_idx(inner_idx).await),
+                None => None,
+            }
+        }
+    }
+}
+
+impl<T: EpochMapper> IndexMapper for T {
+    async fn to_inner_idx(&self, outer_idx: OuterIdx) -> InnerIdx {
+        InnerIdx(
+            self.to_incremental_epoch(outer_idx.0 as UserEpoch)
+                .await
+                .try_into()
+                .unwrap(),
+        )
+    }
+
+    async fn to_outer_idx(&self, inner_idx: InnerIdx) -> OuterIdx {
+        OuterIdx(self.to_user_epoch(inner_idx.0 as IncrementalEpoch).await as usize)
+    }
+}
+
+impl IndexMapper for State {
+    async fn to_inner_idx(&self, outer_idx: OuterIdx) -> InnerIdx {
+        self.inner_idx(outer_idx)
+    }
+
+    async fn to_outer_idx(&self, inner_idx: InnerIdx) -> OuterIdx {
+        self.outer_idx(inner_idx)
+    }
 }
 
 #[derive(Default)]
-pub struct Tree;
-impl Tree {
+pub struct Tree<const IS_EPOCH_TREE: bool>;
```

Comment:
Could we replace `IS_EPOCH_TREE` with something more generic?

Reply:
Do you have any suggestions? Couldn't find a more proper name as the explicit handling of epochs would happen only in case this tree is used to store epoch numbers

Reply:
`IS_SEQUENTIAL_TREE`?

Reply:
This flag is not related to whether epochs are sequential or not, it is only employed to understand whether we should update the epoch mapper of the storage or not, and the epoch mapper is used even with sequential epochs. So I am not sure that `IS_SEQUENTIAL_TREE` would be a good name for this. Maybe something like `HAS_EPOCH_MAPPER` would be better?